metaflow/plugins/kfp/kfp_step_function.py

```diff
@@ -102,3 +356,86 @@ def kfp_step_function(
         "StepOpRet", ["foreach_splits"] + list(preceding_component_inputs_dict.keys())
     )(*values)
     return ret
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--metaflow_run_id", type=str, required=True)
+    parser.add_argument("--metaflow_configs", type=json.loads, required=True)
+    parser.add_argument("--cd_into_metaflow_package_cmd", type=str, required=True)
+    parser.add_argument("--clean_volume_cmd", type=str, required=True)
+    parser.add_argument("--task_id", type=str, required=True)
+    parser.add_argument("--task_id_template", type=str, required=True)
+    parser.add_argument("--step_name", type=str, required=True)
+    parser.add_argument("--flow_name", type=str, required=True)
+    parser.add_argument("--namespace", action="store_true")
+    parser.add_argument("--tags", type=json.loads, required=True)
+    parser.add_argument("--need_split_index", action="store_true")
+    parser.add_argument("--environment_type", type=str, required=True)
+    parser.add_argument("--logger_type", type=str, required=True)
+    parser.add_argument("--monitor_type", type=str, required=True)
+    parser.add_argument("--user_code_retries", type=int, required=True)
+    parser.add_argument("--workflow_name", type=str, required=True)
+    parser.add_argument("--script_name", type=str, required=True)
+    parser.add_argument("--passed_in_split_indexes", type=str, required=False)
+    parser.add_argument("--preceding_component_inputs", type=json.loads, required=False)
+    parser.add_argument(
+        "--preceding_component_outputs", type=json.loads, required=False
+    )
+    parser.add_argument("--flow_parameters_json", type=str, required=False)
+
+    # parse_known_args parses arguments specified above into args, and returns
+    # the rest as a list. This allows us
+    # to pass a dictionary of type [str, dsl.PipelineParam] without serialization issues.
+    args, preceding_component_outputs_dict_args = parser.parse_known_args()
+
+    # Obtain the variable names and values from preceding_component_outputs_dict.
+    # We pass a string in kfp.py with the keys and values in preceding_component_outputs_dict
+    # separated by `=`, e.g. `--dividend=3 --divisor=4`.
+    kwargs = {}
+    for arg in preceding_component_outputs_dict_args:
+        key, value = arg.split("=")
+        kwargs[key] = value
+
+    # We replicate what is done in the KFP SDK _container_op.py,
+    # see: https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/dsl/_container_op.py
+    # We write outputs to a tmp file, which KFP internally uses to produces the output
+    # of the container op.
+    _output_files = ["/tmp/outputs/foreach_splits/data"]
+    for preceding_component_input in args.preceding_component_inputs:
+        _output_files.append(f"/tmp/outputs/{preceding_component_input}/data")
```

Comment:
* Why is `_output_files` declared here then call `kfp_step_function` then write them out? If there's no reason, they. like code should stay together
* We write `_output_files` even if we are not in a foreach split?

Reply:
`output_file` writing is done within `kfp_step_function` and we not longer write if we are not in a foreach split (unless there are preceding kfp components).